### PR TITLE
Change quantitative properties to add units to name

### PR DESCRIPTION
**Pull request by alessandratrapani**
Change quantitative properties from attributes to datasets and add `unit` property as an attribute of each dataset that is numerical.
Additional adjustments:
- [x] for datasets: change required:false --> quantity:'?'
- [x] remove unnecessary class definition from patterned_ogen.py
- [x] unit property would not be modifiable 
- [x] add dims and shape
- [x] add unit measure in the name of the attribute (https://github.com/NeurodataWithoutBorders/nwb-schema/issues/569), instead as defined them a (scalar) dataset + unit-attribute

### CODE REVIEW

**Comment by CodyCBakerPhD** [comment]
Some conflicts here now

**Comment by CodyCBakerPhD** [comment]
The main reasoning for this was to be able to set a `unit` attribute then on each quantitative dataset, right?

Would you be able to share an example file using this alteration to see how it looks in practice?

**Comment by alessandratrapani** [comment]
> The main reasoning for this was to be able to set a `unit` attribute then on each quantitative dataset, right?

Yes, sorry it's still work in progress. I will have a meeting with Ryan tomorrow to solve this

**Comment by alessandratrapani** [comment]
Here is an [example file](https://catalystneuro-processing.s3.us-east-2.amazonaws.com/adesnik-lab-to-nwb/basics_tutorial_patterned_ogen.nwb).
Not quite sure why tests are failing here, but not locally.

**Comment by CodyCBakerPhD** [comment]
> Not quite sure why tests are failing here, but not locally.

Did you try a fresh environment that is set up in exactly the same way the CI does? We as developers can often do little things differently on our own systems

File: spec/ndx-patterned-ogen.extensions.yaml [hunk 1]
```diff
@@ -6,98 +6,140 @@ groups:
...
+        dims:
+        - diameter
+        - width, height
+        shape:
+        - - 1
+        - - 2
+        attributes:
+        - name: unit
+          value: micrometers
+          doc: unit of measure of the scanning sweep pattern size, fixed to micrometers
```
Comment (CodyCBakerPhD) [inline spec/ndx-patterned-ogen.extensions.yaml]:
```suggestion
          doc: Unit of measure of the scanning sweep pattern size, fixed to micrometers.
```

Minor

File: spec/ndx-patterned-ogen.extensions.yaml [hunk 2]
```diff
@@ -6,98 +6,140 @@ groups:
...
+        dims:
+        - diameter
+        - width, height, depth
+        shape:
+        - - 1
+        - - 3
+        attributes:
+        - name: unit
+          value: micrometers
+          doc: unit of measure of the scanning sweep pattern size, fixed to micrometers
```
Comment (CodyCBakerPhD) [inline spec/ndx-patterned-ogen.extensions.yaml]:
```suggestion
          doc: Unit of measure of the scanning sweep pattern size, fixed to micrometers.
```

File: spec/ndx-patterned-ogen.extensions.yaml [hunk 3]
```diff
@@ -6,98 +6,140 @@ groups:
...
-    attributes:
+    datasets:
       - name: diameter
         dtype: numeric
         doc: Spiral diameter (in micrometers).
-        required: true
+        attributes:
+        - name: unit
+          value: micrometers
+          doc: unit of measure of the spiral diameter, fixed to micrometers
```
Comment (CodyCBakerPhD) [inline spec/ndx-patterned-ogen.extensions.yaml]:
```suggestion
          doc: Unit of measure of the spiral diameter, fixed to micrometers.
```

File: spec/ndx-patterned-ogen.extensions.yaml [hunk 4]
```diff
@@ -6,98 +6,140 @@ groups:
...
-        doc: Describe any additional details about the pattern.
-        required: false
       - name: height
         dtype: numeric
         doc: Spiral height of each sweep (in micrometers).
+        quantity: "?"
+        attributes:
+        - name: unit
+          value: micrometers
+          doc: unit of measure of the spiral height, fixed to micrometers
```
Comment (CodyCBakerPhD) [inline spec/ndx-patterned-ogen.extensions.yaml]:
```suggestion
          doc: Unit of measure of the spiral height, fixed to micrometers.
```

File: spec/ndx-patterned-ogen.extensions.yaml [hunk 5]
```diff
@@ -122,45 +169,80 @@ groups:
...
         doc: Excitation wavelength of stimulation light (nanometers).
-        required: true
-      - name: model
-        dtype: text
-        doc: Model of light source device.
-        required: false
+        attributes:
+        - name: unit
+          value: nanometers
+          doc: unit of measure of wavelength which is fixed to nanometers
```
Comment (CodyCBakerPhD) [inline spec/ndx-patterned-ogen.extensions.yaml]:
```suggestion
          doc: Unit of measure of wavelength, fixed to nanometers.
```

Also adjusted to be consistent with above

File: spec/ndx-patterned-ogen.extensions.yaml [hunk 6]
```diff
@@ -122,45 +169,80 @@ groups:
...
+        quantity: "?"
       - name: peak_power
         dtype: numeric
         doc: Incident power of stimulation device (in Watts).
-        required: false
+        quantity: "?"
+        attributes:
+        - name: unit
+          value: Watts
+          doc: unit of measure of power which is fixed to Watts
```
Comment (CodyCBakerPhD) [inline spec/ndx-patterned-ogen.extensions.yaml]:
```suggestion
          doc: Unit of measure of power, fixed to Watts.
```

File: spec/ndx-patterned-ogen.extensions.yaml [hunk 7]
```diff
@@ -122,45 +169,80 @@ groups:
...
+          dtype: text
       - name: peak_pulse_energy
         dtype: numeric
         doc: If device is pulsed light source, pulse energy (in Joules).
-        required: false
+        attributes:
+        - name: unit
+          value: Joules
+          doc: unit of measure of energy which is fixed to Joules
```
Comment (CodyCBakerPhD) [inline spec/ndx-patterned-ogen.extensions.yaml]:
```suggestion
          doc: Unit of measure of energy, fixed to Joules.
```

File: spec/ndx-patterned-ogen.extensions.yaml [hunk 8]
```diff
@@ -122,45 +169,80 @@ groups:
...
+        quantity: "?"
       - name: intensity
         dtype: numeric
         doc: Intensity of the excitation in W/m^2, if known.
-        required: false
+        quantity: "?"
+        attributes:
+        - name: unit
+          value: W/m^2
+          doc: unit of measure of intensity which is fixed to Watts per square meters
```
Comment (CodyCBakerPhD) [inline spec/ndx-patterned-ogen.extensions.yaml]:
```suggestion
          doc: Unit of measure of intensity, fixed to Watts per square meters.
```

File: spec/ndx-patterned-ogen.extensions.yaml [hunk 9]
```diff
@@ -122,45 +169,80 @@ groups:
...
+          dtype: text        
       - name: exposure_time
         dtype: numeric
         doc: Exposure time of the sample (in sec).
-        required: false
+        quantity: "?"
+        attributes:
+        - name: unit
+          value: seconds
+          doc: unit of measure of exposure time which is fixed to seconds
```
Comment (CodyCBakerPhD) [inline spec/ndx-patterned-ogen.extensions.yaml]:
```suggestion
          doc: Unit of measure of exposure time, fixed to seconds.
```

File: spec/ndx-patterned-ogen.extensions.yaml [hunk 10]
```diff
@@ -122,45 +169,80 @@ groups:
...
+          dtype: text   
       - name: pulse_rate
         dtype: numeric
         doc: If device is pulsed light source, pulse rate (in Hz) used for stimulation.
+        quantity: "?"
+        attributes:
+        - name: unit
+          value: Hz
+          doc: unit of measure of rate which is fixed to Hertz
```
Comment (CodyCBakerPhD) [inline spec/ndx-patterned-ogen.extensions.yaml]:
```suggestion
          doc: Unit of measure of rate, fixed to Hertz.
```

File: spec/ndx-patterned-ogen.extensions.yaml [hunk 11]
```diff
@@ -169,46 +251,55 @@ groups:
...
     datasets:
       - name: power
         neurodata_type_inc: VectorData
-        doc: Power (in Watts) applied to each target during patterned photostimulation.
-        required: true
+        doc: Power applied to each target during patterned photostimulation, in W.
+        attributes:
+        - name: unit
+          value: Watts
+          doc: unit of measure of power which is fixed to Watts
```
Comment (CodyCBakerPhD) [inline spec/ndx-patterned-ogen.extensions.yaml]:
```suggestion
          doc: Unit of measure of power, fixed to Watts.
```

File: spec/ndx-patterned-ogen.extensions.yaml [hunk 12]
```diff
@@ -169,46 +251,55 @@ groups:
...
       - name: frequency
         neurodata_type_inc: VectorData
-        doc: Frequency of stimulation if the stimulus delivered is pulsed (in Hz).
-        required: false
+        doc: Frequency of stimulation if the stimulus delivered is pulsed, in Hz.
+        quantity: "?"
+        attributes:
+        - name: unit
+          value: Hertz
+          doc: unit of measure of frequency which is fixed to Hertz
```
Comment (CodyCBakerPhD) [inline spec/ndx-patterned-ogen.extensions.yaml]:
```suggestion
          doc: Unit of measure of frequency, fixed to Hertz.
```

File: spec/ndx-patterned-ogen.extensions.yaml [hunk 13]
```diff
@@ -169,46 +251,55 @@ groups:
...
+          dtype: text
       - name: pulse_width
         neurodata_type_inc: VectorData
         doc: Pulse width of stimulation if the stimulus delivered is pulsed, in seconds/phase.
-        required: false
+        quantity: "?"
+        attributes:
+        - name: unit
+          value: seconds/phase
+          doc: unit of measure of power which is fixed to seconds/phase
```
Comment (CodyCBakerPhD) [inline spec/ndx-patterned-ogen.extensions.yaml]:
```suggestion
          doc: Unit of measure of power, fixed to seconds per phase.
```

I like the approach where the doc expands the symbolic representations of the complex SI form in pure english

**Comment by CodyCBakerPhD** [comment]
It's weird; the shared file has an `excitation_lambda` field as a part of its `PatternedOptogeneticStimulusSite` but I don't see that field specified in the schema? Any ideas (is it purely inherited)? I just notice because it doesn't have a unit like the others do

**Comment by CodyCBakerPhD** [comment]
@alessandratrapani Thanks a bunch for working on this, it's great to see an example in practice

Just something to leave as a note here; adding lots of tiny datasets to the file may affect standard performance of read operations, especially from the cloud. Which might diminish some of the gains we've made with the latest version of the extension overall

I'm banking on us figuring out better ways to read from the cloud (kerchunk/consolidated metadata) to resolve that, but good to be aware of and consider if the gain in specificity of the metadata values is worth it

**Comment by alessandratrapani** [comment]
> It's weird; the shared file has an `excitation_lambda` field as a part of its `PatternedOptogeneticStimulusSite` but I don't see that field specified in the schema? Any ideas (is it purely inherited)? I just notice because it doesn't have a unit like the others do

Yes, it's purely inherited.

**Comment by CodyCBakerPhD** [comment]
@alessandratrapani Is there a stub file with the new attribute name change approach instead of datasets?

**Comment by alessandratrapani** [comment]
> @alessandratrapani Is there a stub file with the new attribute name change approach instead of datasets?

yes, sorry I forgot to update it: [here](https://catalystneuro-processing.s3.us-east-2.amazonaws.com/adesnik-lab-to-nwb/basics_tutorial_patterned_ogen.nwb)

**Review by CodyCBakerPhD — APPROVED** [review]

**Comment by CodyCBakerPhD** [comment]
I think this looks great this way

**Pull request merged**